### PR TITLE
Filled in the empty data types and completed functions

### PR DESCRIPTION
**Pull request by mamboelrey**
having a trouble with linker error when compiles

### CODE REVIEW

File: main.cpp [hunk 1]
```diff
@@ -54,48 +54,59 @@ Purpose:  This project will show you the difference between member functions and
 
 #include <iostream>
 #include <string>
-struct T
+struct SoccerBall
 {
-    T(<#type name#> v, const char* <#variable name#>)   //1
-    //2
+    SoccerBall(float v, const char* brand);   //1
+    //2
```
Comment (matkatmusic) [inline main.cpp]: Where is the rest of this function?
- [x] revise

File: main.cpp [hunk 2]
```diff
@@ -54,48 +54,59 @@ Purpose:  This project will show you the difference between member functions and
...
 
-struct <#structName1#>                                //4
+struct SoccerBallSize                               //4
 {
-    <#type name#> compare(<#type name#> a, <#type name#> b) //5
+    SoccerBall* compare(SoccerBall* a, SoccerBall* b) //5
     {
         if( a->value < b->value ) return a;
         if( a->value > b->value ) return b;
         return nullptr;
```
Comment (matkatmusic) [inline main.cpp]: What is the one thing we always do before we use pointers?
- [x] revise

File: main.cpp [hunk 3]
```diff
@@ -54,48 +54,59 @@ Purpose:  This project will show you the difference between member functions and
...
 {
-    float <#name1#> { 0 }, <#name2#> { 0 };
-    <#returnType#> <#memberFunction#>(<#type name#>* <#updatedValue#>)      //12
+    float danielBall { 0.0f }, anthonyBall { 0.0f };
+    float decrementBallDistance(float* updatedValue)      //12
     {
-        
+        std::cout << "StartDistance's danielBall value: " << danielBall << std::endl;
+        danielBall = *updatedValue;
+        std::cout << "StartDistance's danielBall updated value: " << danielBall << std::endl;
```
Comment (matkatmusic) [inline main.cpp]: What is the one thing we always do before we use pointers?
- [x] revise

File: main.cpp [hunk 4]
```diff
@@ -54,48 +54,59 @@ Purpose:  This project will show you the difference between member functions and
...
+    static int kickSoccerBall(StartDistance* that, float* updatedValue)        //10
     {
-        std::cout << "U's <#name1#> value: " << that-><#name1#> << std::endl;
-        that-><#name1#> = <#updatedValue#>;
-        std::cout << "U's <#name1#> updated value: " << that-><#name1#> << std::endl;
-        while( std::abs(that-><#name2#> - that-><#name1#>) > 0.001f )
+        std::cout << "StartDistance's danielBall value: " << that->danielBall << std::endl;
+        that->danielBall = *updatedValue;
+        std::cout << "StartDistance's danielBall updated value: " << that->danielBall << std::endl;
+        while( std::abs(that->anthonyBall - that->danielBall) > 0.001f )
```
Comment (matkatmusic) [inline main.cpp]: What is the one thing we always do before we use pointers?
- [x] revise

File: main.cpp [hunk 5]
```diff
@@ -115,17 +126,21 @@ struct <#structname2#>
...
+    SoccerBall walmartBall( 4.4f, "Wilson");                                             //6
+    SoccerBall targetBall(5.5f, "Nike");                                             //6
     
-    <#structName1#> f;                                            //7
-    auto* smaller = f.compare( , );                              //8
+    SoccerBallSize f;                                            //7
+    auto* smaller = f.compare(&walmartBall, &targetBall);                              //8
+    if(smaller == nullptr)
+        return -1;
     std::cout << "the smaller one is << " << smaller->name << std::endl; //9
```
Comment (matkatmusic) [inline main.cpp]: Don't return prematurely. 
Instead, print out the statement if `smaller` is NOT a nullptr
- [x] revise

**Review by matkatmusic — CHANGES_REQUESTED** [review]

File: main.cpp [hunk 6]
```diff
@@ -54,48 +54,68 @@ Purpose:  This project will show you the difference between member functions and
...
+        return anthonyBall * danielBall;
     }
 };
 
-struct <#structname2#>
+struct KickDistance
 {
-    static <#returntype#> <#staticFunctionA#>(U* that, <#type name#>* <#updatedValue#> )        //10
+    static int kickSoccerBall(StartDistance* that, float* updatedValue)        //10
     {
```
Comment (matkatmusic) [inline main.cpp]: This code produces a warning
- [x] revise

**Review by matkatmusic — CHANGES_REQUESTED** [review]

File: main.cpp [hunk 7]
```diff
@@ -54,48 +54,68 @@ Purpose:  This project will show you the difference between member functions and
...
 {
-    <#type name#> compare(<#type name#> a, <#type name#> b) //5
+    SoccerBall* compare(SoccerBall* a, SoccerBall* b) //5
     {
-        if( a->value < b->value ) return a;
-        if( a->value > b->value ) return b;
+        if( a->value < b->value && a != nullptr && b != nullptr ) return a;
+        if( a->value > b->value && a != nullptr && b != nullptr ) return b;
         return nullptr;
     }
```
Comment (matkatmusic) [inline main.cpp]: You need to check if `a` and `b` are not nullptrs first before you try to use them.
Expressions are evaluated from left to right.  Your code is using `a` and `b` before checking if it's ok to use them.
```
if( a != nullptr && b != nullptr )
{
    //it's ok to use them
}

return nullptr;
```
- [x] revise

File: main.cpp [hunk 8]
```diff
@@ -54,48 +54,68 @@ Purpose:  This project will show you the difference between member functions and
...
-        that-><#name1#> = <#updatedValue#>;
-        std::cout << "U's <#name1#> updated value: " << that-><#name1#> << std::endl;
-        while( std::abs(that-><#name2#> - that-><#name1#>) > 0.001f )
+        if(updatedValue != nullptr)
+        {
+            std::cout << "StartDistance's danielBall value: " << that->danielBall << std::endl;
+        }
+        if(that != nullptr)
+            that->danielBall = *updatedValue;
+        std::cout << "StartDistance's danielBall updated value: " << that->danielBall << std::endl;
```
Comment (matkatmusic) [inline main.cpp]: You are still using `updatedValue` even though it might be a nullptr.
- [x] revise

**Review by matkatmusic — CHANGES_REQUESTED** [review]

File: main.cpp [hunk 9]
```diff
@@ -54,48 +54,76 @@ Purpose:  This project will show you the difference between member functions and
...
-struct T
+struct SoccerBall
 {
-    T(<#type name#> v, const char* <#variable name#>)   //1
-    //2
+    SoccerBall(float v, const char* brand): value(0), name("")
+    {
+        value = v;
+        name = brand;
+    }   //1
```
Comment (matkatmusic) [inline main.cpp]: Use the member initializer list correctly.
pass the constructor arguments to the member variables
- [x] revise

File: main.cpp [hunk 10]
```diff
@@ -113,19 +141,31 @@ struct <#structname2#>
...
+        
+        StartDistance ballStart1;
+        float updatedValue = 5.0f;
+        std::cout << "kickSoccerBall ballStart's multiplied values: " << KickDistance::kickSoccerBall(&ballStart1, &updatedValue) << std::endl;                  //11
+        
+        StartDistance ballStart2;
+        std::cout << "kickSoccerBall continuousDistance's multiplied values: " << ballStart2.decrementBallDistance(&updatedValue) << std::endl;
+
+        return 0;
+    }
```
Comment (matkatmusic) [inline main.cpp]: All of this stuff happens regardless of the value stored in `smaller`
It is not part of the `if(smaller != nullptr)` block
- [x] revise

**Review by matkatmusic — CHANGES_REQUESTED** [review]
you made some odd programming choices in the latest revision.

File: main.cpp (hunk 9 above)
Comment (mamboelrey) [reply main.cpp]: Is this odd because I am using the constructor initializer list to values other than the parameters passed into the constructor?

**Review by matkatmusic — APPROVED** [review]

**Pull request merged**